(dev/core#5665) CiviTest/bootstrap.php - Don't try to init DB before tests

totten opened the pull request:
Overview
----------------------------------------

Remove an extraneous call that initializes the database. The essential change is here (5de7bff4e7e06a277381e1aee0083d8591905633), and everything else is dealing with edge-cases that had some implicit reliance.

Before
----------------------------------------

When you run a test, it configures the database twice.

```
CIVICRM_UF=UnitTests phpunit9 tests/phpunit/CRM/Core/RegionTest.php --filter testBlank
```
```
Installing dmastertest_hskvs schema                              <<=== FIRST
PHPUnit 9.6.20 by Sebastian Bergmann and contributors.


Installing dmastertest_hskvs database                            <<=== SECOND
.                                                                   1 / 1 (100%)

Time: 00:02.688, Memory: 101.88 MB

OK (1 test, 7 assertions)
```


After
----------------------------------------

When you run a test, it configures the database once.

```
CIVICRM_UF=UnitTests phpunit9 tests/phpunit/CRM/Core/RegionTest.php --filter testBlank
```
```
PHPUnit 9.6.20 by Sebastian Bergmann and contributors.


Initializing "Headless System" (1d4f6e2b0d21d118dac28b3a24f10585) in "dmastertest_hskvs"
.                                                                   1 / 1 (100%)

Time: 00:04.807, Memory: 101.88 MB

OK (1 test, 7 assertions)
```

totten committed:
CiviTest/bootstrap.php - Don't try to init DB before tests

eileenmcnaughton commented:
there is a weird one in there - https://github.com/civicrm/civicrm-core/pull/31685

totten committed:
APIv4 ConformanceTest - Fix getEntitiesLotech()

totten committed:
CiviUnitTestCase - Ensure init on empty DB

totten commented:
Yeah, the Smarty errors are weird.

I *think* the problem is arising on the first test of the suite. Reading the call-stack of that test more carefully, these stand out:

```
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/BAO/Domain.php:64
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/I18n.php:618
...
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/Smarty.php:244
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/Smarty/UserContentPolicy.php:192
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/Smarty/plugins/function.crmSqlData.php:30
...
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Template.php:125
/home/...build-4/web/sites/all/modules/civicrm/CRM/Core/CodeGen/PhpSchema.php:98
/home/...build-4/web/sites/all/modules/civicrm/Civi/Test/Data.php:26
```

So it's in a pre-install environment, generating SQL, calling Smarty, and then Smarty is trying to access higher-level services (like `Domain`s) which aren't available. 

* That's not supposed to happen because `CRM_Core_CodeGen_Util_Template` uses a smaller/private instance of Smarty... but the add-on `{crmSqlData}` goes to the full-Smarty.
* I'm not sure why the patch here causes a change between pass<=>fail, but the failure seems fair.
* The failure happens while *initializing* full-Smarty... so then that instance might be corrupted/incomplete. And that might explain why subsequent tests fail.

totten committed:
crmSqlData - Fix compat with pre-install environment

eileenmcnaughton commented:
That's complicated!

I'm gonna chip away at some of the tests on the other path - ie calling it less often in `setUpBeforeClass()` cos it feels like if we can sort that one out it will give some real performance advantage - this one feels more marginal in terms of gain

totten committed:
(NFC) CiviUnitTestCase - Comments

totten commented:
@eileenmcnaughton This is passing now...

totten committed:
APIv3 SyntaxConformanceTest - Fix dynamic providers

totten committed:
Civi\Test - Consolidate messaging

totten commented:
Rebased. Added description. Tweaked output.

Also, if there are more dynamic-data-providers that haven't bubbled up yet, a simple way to fix them is to put the `apply()` call in the provider. As in:

```diff
+++ Patch for dynamic data-provider in an instance of CiviUnitTestCase
 public static function myDataProvider(): array {
+  \Civi\Test::asPreInstall([static::CLASS, 'buildEnvironment'])->apply();
   $dynamic = CallLots::ofMethods()->and('stuff');
   return $dynamic;
 }
```

or

```diff
+++ Patch for dynamic data-provider in an instance of HeadlessInterface
 public static function myDataProvider(): array {
+  (new static())->setUpHeadless();
   $dynamic = CallLots::ofMethods()->and('stuff');
   return $dynamic;
 }
```
